sound/soc/sof/ipc4-topology.c

```diff
@@ -1451,6 +1452,21 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
 		msg = &pipeline->msg;
 		msg->primary |= pipeline->mem_usage;
+
+		if (ipc4_data->max_num_pipelines <= 0) {
+			dev_err(sdev->dev, "Invalid max_num_pipelines %d\n",
+				ipc4_data->max_num_pipelines);
+			return ipc4_data->max_num_pipelines;
+		}
+		swidget->instance_id = ida_alloc_max(&ppl_ida, ipc4_data->max_num_pipelines,
+						     GFP_KERNEL);
```

Comment:
XArray is the preferred over the deprecated IDA...

Reply:
You might want to free the allocation in the error cases after this point.

Reply:
Thanks @ujfalusi I added `ida_free(&pipeline_ida, swidget->instance_id);` in error. But, could you describe the XArray comment in more detail?  Which part do you prefer using XArray? Thanks

Reply:
is this correct @ujfalusi that ida_ is deprecated? That's is true of *idr*, 

https://docs.kernel.org/next/core-api/idr.html?highlight=idr#idr-usage

````
The IDR interface is deprecated; please use the [XArray](https://docs.kernel.org/next/core-api/xarray.html) instead.
````
but that's a different API.

What makes you think the IDA is deprecated? We clearly just added it for all kinds of things recently.

Reply:
True, sorry